In [1]:
import pandas as pd

import warnings
warnings.filterwarnings('ignore')

## Notes:
- 2015 regular season 1st and 2nd were also playoff 1st and 2nd
- 2016 playoff winners were 5th seed (1st) and 3rd seed (2nd)
- 2017 regular season 1st and 2nd were also playoff 1st and 2nd
- 2018 playoff winners were 5th seed (1st) and 2nd seed

## 2015 - Regular Season

In [2]:
twenty15_reg = pd.ExcelFile('2015.xlsx')
df_2015_reg_1st = pd.read_excel(twenty15_reg, '1st')
df_2015_reg_2nd = pd.read_excel(twenty15_reg, '2nd')

df_2015_po_1st = df_2015_reg_1st
df_2015_po_2nd = df_2015_reg_2nd

In [3]:
df_2015_reg_1st

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0        Nicolas Batum       SF   372/873  0.426  163/192  0.849   139  1046   
1        Avery Bradley       SG  456/1019  0.447   96/123  0.780   147  1155   
2     Robert Covington       SF   274/712  0.385  140/177  0.791   170   858   
3        Jordan Farmar       PG    42/100  0.420    10/10  1.000    16   110   
4       Derrick Favors       PF   418/811  0.515  180/254  0.709     0  1016   
5          Danny Green       SF   211/561  0.376    34/46  0.739   116   572   
6           Al Horford        C  529/1048  0.505  103/129  0.798    88  1249   
7       Victor Oladipo       SG   425/970  0.438  205/247  0.830    98  1153   
8        Elfrid Payton       PG   318/730  0.436  112/190  0.589    30   778   
9      Otto Porter Jr.       SF   339/717  0.473   95/126  0.754    98   871   
10     Josh Richardson       SF   123/272  0.452    42/63  0.667    53   341   
11  Karl-Anthony Towns        C  625/1153  0.542  223/275  0.811    30  1503   
12        Kemba Walker       PG  568/1331  0.427  371/438  0.847   182  1689   

    REB  AST   ST  BLK   TO  auction_value  
0   428  403   65   43  205             17  
1   220  158  117   19  109              0  
2   419   97  105   40  143              8  
3    25   37   15    2   18              0  
4   503   94   74   91  104             16  
5   303  141   79   64   75             10  
6   596  263   68  121  107             36  
7   345  282  116   54  153             21  
8   261  467   89   20  177              6  
9   391  120  105   32   70              6  
10  107   73   36   25   35              0  
11  857  161   58  138  183             12  
12  357  421  126   39  171             14

In [4]:
df_2015_reg_2nd

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0       Trevor Ariza       SF   357/859  0.416  126/161  0.783   185  1025   
1        Jae Crowder       SF   360/813  0.443  196/239  0.820   122  1038   
2      Toney Douglas       SG   183/445  0.411    84/99  0.848    81   531   
3          Pau Gasol        C   467/995  0.469  229/289  0.792    24  1187   
4        Rudy Gobert        C   198/354  0.559  161/283  0.569     0   557   
5        George Hill       PG   326/739  0.441  114/150  0.760   128   894   
6       Shane Larkin       PG   222/502  0.442   83/107  0.776    39   566   
7         Kyle Lowry       PG  512/1198  0.427  398/491  0.811   212  1634   
8        Ian Mahinmi        C   264/448  0.589  132/225  0.587     0   660   
9    Wesley Matthews       SG   331/854  0.388  126/146  0.863   189   977   
10   Thomas Robinson       PF   130/291  0.447   47/109  0.431     0   307   
11         Ish Smith       PG   397/967  0.411  122/176  0.693    56   972   
12  Hassan Whiteside        C   412/681  0.605  214/329  0.650     0  1038   

    REB  AST   ST  BLK   TO  auction_value  
0   366  188  160   26  113             18  
1   373  135  126   35   83              2  
2   143  157   70    5   64              0  
3   793  294   42  146  164             32  
4   668   91   45  135  114             30  
5   298  259   83   17  101              4  
6   180  343   95   12  146              0  
7   365  494  158   34  225             35  
8   507  104   65   75  100              0  
9   238  151   78   17   78              7  
10  359   46   39   36   69              0  
11  305  502   88   24  175              0  
12  865   30   44  269  137             25

In [5]:
print("Total value of 1st-place team:", df_2015_reg_1st['auction_value'].sum())
print("Total value of 2nd-place team:", df_2015_reg_2nd['auction_value'].sum())

Total value of 1st-place team: 146
Total value of 2nd-place team: 153


### 2015 Recap

- In 2015, the top two teams in the regular season were also the winners in the playoffs.
- No players selected for over $36 (18%) of a team's budget

## 2016 - Regular Season

In [6]:
twenty16_reg = pd.ExcelFile('2016_reg.xlsx')
df_2016_reg_1st = pd.read_excel(twenty16_reg, '1st')
df_2016_reg_2nd = pd.read_excel(twenty16_reg, '2nd')

In [7]:
df_2016_reg_1st

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0      Carmelo Anthony       SF  602/1389  0.433  304/365  0.833   151  1659   
1      Harrison Barnes       SF  599/1280  0.468  242/281  0.861    78  1518   
2         Eric Bledsoe       PG  449/1034  0.434  388/458  0.847   104  1390   
3      Malcolm Brogdon       PG   290/635  0.457  109/126  0.865    78   767   
4     DeMarcus Cousins        C  647/1432  0.452  517/670  0.772   131  1942   
5         Jrue Holiday       PG   405/893  0.454  119/168  0.708   100  1029   
6       Frank Kaminsky        C   320/802  0.399  118/156  0.756   116   874   
7           Jeremy Lin       PG   175/400  0.438  115/141  0.816    58   523   
8      Larry Nance Jr.       PF   190/361  0.526    59/80  0.738    10   449   
9         Nerlens Noel        C   184/309  0.595   77/111  0.694     0   445   
10  Kristaps Porzingis        C   443/985  0.450  198/252  0.786   112  1196   
11       Isaiah Thomas       PG  682/1473  0.463  590/649  0.909   245  2199   
12   Jonas Valanciunas        C   391/702  0.557  176/217  0.811     1   959   

    REB  AST   ST  BLK   TO  auction_value  
0   437  213   61   33  153             27  
1   394  117   65   15  102              6  
2   320  418   91   31  224             22  
3   213  317   84   12  113              0  
4   794  332  100   94  269             52  
5   264  488  100   44  194              1  
6   337  162   47   34   76              0  
7   135  184   42   15   86             17  
8   369   96   82   40   56              1  
9   294   50   65   50   51              6  
10  475   97   47  129  118             25  
11  205  448   70   13  210             32  
12  759   57   37   63  106             12

In [8]:
df_2016_reg_2nd

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0      Trevor Ariza       SF   326/798  0.409   93/126  0.738   191   936   
1   Wilson Chandler       SF   433/939  0.461  141/194  0.727   110  1117   
2     Anthony Davis        C  770/1526  0.505  519/647  0.802    40  2099   
3      Goran Dragic       PG  534/1124  0.475  298/377  0.790   117  1483   
4      Kevin Durant       SF  551/1026  0.537  336/384  0.875   117  1555   
5      Yogi Ferrell       PG   160/394  0.406    74/89  0.831    68   462   
6        Marc Gasol        C  532/1160  0.459  278/332  0.837   104  1446   
7       Joe Johnson       SF   273/625  0.437    63/77  0.818   106   715   
8     Dirk Nowitzki       PF   296/678  0.437   98/112  0.875    79   769   
9   Josh Richardson       SF   202/513  0.394    60/77  0.779    75   539   
10      Ricky Rubio       PG   261/650  0.402  254/285  0.891    60   836   
11       Tyler Ulis       PG   184/437  0.421    55/71  0.775    21   444   
12      Ivica Zubac        C   126/238  0.529    32/49  0.653     0   284   

    REB  AST   ST  BLK   TO  auction_value  
0   458  175  147   20   73             14  
1   460  141   51   30  114              1  
2   884  157   94  167  181             53  
3   279  424   89   13  212             11  
4   513  300   66   99  138             66  
5   111  172   42    9   70              0  
6   463  338   67   99  166             17  
7   243  144   36   18   70              0  
8   353   82   30   38   51              8  
9   169  139   60   40   65              0  
10  305  682  128   10  195             10  
11   95  226   48    5   77              0  
12  159   30   14   33   30              0

In [9]:
print("Total value of 1st-place team:", df_2016_reg_1st['auction_value'].sum())
print("Total value of 2nd-place team:", df_2016_reg_2nd['auction_value'].sum())

Total value of 1st-place team: 201
Total value of 2nd-place team: 180


## 2016 - Playoff Teams

In [10]:
twenty16_po = pd.ExcelFile('2016_playoff.xlsx')
df_2016_po_1st = pd.read_excel(twenty16_po, '1st')
df_2016_po_2nd = pd.read_excel(twenty16_po, '2nd')

In [11]:
df_2016_po_1st

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0   LaMarcus Aldridge        C  500/1049  0.477  220/271  0.812    23  1243   
1       Avery Bradley       SG   359/775  0.463    68/93  0.731   108   894   
2        Jimmy Butler       SF  570/1252  0.455  585/676  0.865    91  1816   
3         Mike Conley       PG  464/1010  0.459  316/368  0.859   171  1415   
4          Seth Curry       PG   338/703  0.481   85/100  0.850   137   898   
5           Luol Deng       SF   164/424  0.387    46/63  0.730    51   425   
6        Gorgui Dieng        C   332/661  0.502  136/167  0.814    16   816   
7    Tim Hardaway Jr.       SG   415/912  0.455  164/214  0.766   149  1143   
8      Richaun Holmes        C   230/412  0.558   72/103  0.699    27   559   
9       Zaza Pachulia        C   164/307  0.534   98/126  0.778     0   426   
10         Chris Paul       PG   374/785  0.476  232/260  0.892   124  1104   
11     Nikola Vucevic        C  483/1031  0.468  107/160  0.669    23  1096   
12     Thaddeus Young       SF   362/687  0.527    45/86  0.523    45   814   

    REB  AST   ST  BLK   TO  auction_value  
0   523  139   46   88   98             39  
1   334  122   68   11   88              7  
2   470  417  143   32  159             36  
3   242  433   92   19  156             17  
4   179  188   79    7   92              0  
5   295   74   48   20   43              2  
6   647  158   88   95  107             15  
7   224  182   55   15  106              0  
8   311   58   42   54   55              0  
9   410  132   59   33   87              3  
10  303  563  120    8  147             53  
11  779  208   76   74  117             14  
12  448  122  113   30   95             11

In [12]:
df_2016_po_2nd

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0       Al-Farouq Aminu       SF   183/467  0.392   96/136  0.706    70   532   
1         Kent Bazemore       SF   295/721  0.409  119/168  0.708    92   801   
2      Patrick Beverley       PG   228/543  0.420    73/95  0.768   110   639   
3   Willie Cauley-Stein        C   255/481  0.530  101/151  0.669     0   611   
4       Darren Collison       PG   340/714  0.476  147/171  0.860    73   900   
5      Robert Covington       SF   292/732  0.399  143/174  0.822   137   864   
6           Jae Crowder       SF   333/720  0.463  176/217  0.811   157   999   
7         Marcin Gortat        C   390/674  0.579  103/159  0.648     0   883   
8        Damian Lillard       PG  661/1488  0.444  488/545  0.895   214  2024   
9            Kyle Lowry       PG   426/918  0.464  299/365  0.819   193  1344   
10      Markieff Morris       SF   406/889  0.457  180/215  0.837    71  1063   
11         Myles Turner        C   444/869  0.511  245/303  0.809    40  1173   
12         Kemba Walker       PG  643/1449  0.444  304/359  0.847   240  1830   

    REB  AST   ST  BLK   TO  auction_value  
0   451   99   61   44   94              2  
1   231  177   91   52  125             11  
2   393  281   99   25  100              1  
3   338   80   53   44   69              0  
4   152  312   67    9  113              7  
5   436  102  127   69  131              3  
6   415  155   73   23   80              8  
7   849  121   40   61  119              5  
8   368  440   68   20  197             47  
9   286  417   89   19  173             35  
10  493  126   82   43  129              8  
11  589  106   73  173  105             25  
12  308  434   85   22  168             32

In [13]:
print("Total value of 1st-place playoff team:", df_2016_po_1st['auction_value'].sum())
print("Total value of 2nd-place playoff team:", df_2016_po_2nd['auction_value'].sum())

Total value of 1st-place playoff team: 197
Total value of 2nd-place playoff team: 184


## 2017 - Regular Season + Playoffs

In [14]:
twenty17_reg = pd.ExcelFile('2017.xlsx')
df_2017_reg_1st = pd.read_excel(twenty17_reg, '1st')
df_2017_reg_2nd = pd.read_excel(twenty17_reg, '2nd')

df_2017_po_1st = df_2017_reg_1st
df_2017_po_2nd = df_2017_reg_2nd

In [15]:
df_2017_reg_1st

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM  \
0              Jarrett Allen        C   234/397  0.589  114/147  0.776     5   
1   Kentavious Caldwell-Pope       SG   340/798  0.426  153/194  0.789   159   
2                 Quinn Cook       PG   122/252  0.484    22/25  0.880    46   
3             Dewayne Dedmon        C   250/477  0.524    67/86  0.779    50   
4               Goran Dragic       PG  484/1075  0.450  217/271  0.801   111   
5                  Pau Gasol        C   287/627  0.458  158/209  0.756    43   
6               Aaron Gordon       SF   375/865  0.434  157/225  0.698   115   
7             Draymond Green       PF   281/619  0.454  134/173  0.775    77   
8                 Joe Ingles       Sf   335/718  0.467    66/83  0.795   204   
9               Nikola Jokic        C  504/1009  0.500  266/313  0.850   111   
10           Khris Middleton       SF  593/1272  0.466  320/362  0.884   146   
11          Donovan Mitchell       SG  595/1362  0.437  239/297  0.805   187   
12              Kemba Walker       PG  588/1363  0.431  363/420  0.864   231   

     PTS  REB  AST   ST  BLK   TO  auction_value  
0    587  388   49   28   88   82              0  
1    992  387  162  106   16   97              4  
2    312   82   89   12    1   32              0  
3    617  489   90   40   51   86              0  
4   1296  307  359   63   12  166             18  
5    775  619  238   24   79  108              3  
6   1022  457  136   59   45  107              9  
7    773  535  508   95   91  202             38  
8    940  344  392   90   20  159              0  
9   1385  803  458   90   61  210             46  
10  1652  429  328  119   21  191             19  
11  1616  296  291  118   27  215              0  
12  1770  248  444   91   24  179             31

In [16]:
df_2017_reg_2nd

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0       Kyle Anderson       SF   231/438  0.527  104/146  0.712    19   585   
1       Dillon Brooks       SF   340/772  0.440  124/166  0.747    94   898   
2       Anthony Davis        C  780/1462  0.534  495/598  0.828    55  2110   
3      Andre Drummond        C   466/881  0.529  239/395  0.605     0  1171   
4          Marc Gasol        C  434/1033  0.420  281/337  0.834   109  1258   
5      JaMychal Green       PF   223/488  0.457   80/111  0.721    43   569   
6      Nikola Mirotic       PF   306/684  0.447  112/137  0.818   135   859   
7       Dirk Nowitzki        C   346/758  0.456   97/108  0.898   138   927   
8       Elfrid Payton       PG   329/667  0.493  109/168  0.649    30   797   
9    Denzel Valentine       SF   301/721  0.417    38/51  0.745   143   783   
10     Nikola Vucevic        C   399/840  0.475    77/94  0.819    64   939   
11  Russell Westbrook       PG  757/1687  0.449  417/566  0.737    97  2028   
12     Thaddeus Young       SF   421/864  0.487    55/92  0.598    58   955   

     REB  AST   ST  BLK   TO  auction_value  
0    396  202  115   60   94              0  
1    257  135   73   17  124              0  
2    832  174  115  193  162             58  
3   1247  237  114  127  200             16  
4    592  305   54  101  194             29  
5    464   79   32   25   76              1  
6    405   80   44   39   62              1  
7    438  120   43   45   51              2  
8    274  393   83   22  168              7  
9    393  244   65   11   99              0  
10   523  191   56   63  109             17  
11   804  820  147   20  381             66  
12   512  152  135   36  105              4

In [17]:
print("Total value of 1st-place playoff team:", df_2017_reg_1st['auction_value'].sum())
print("Total value of 2nd-place playoff team:", df_2017_reg_2nd['auction_value'].sum())

Total value of 1st-place playoff team: 168
Total value of 2nd-place playoff team: 201


## 2018 - Regular Season

In [18]:
twenty18_reg = pd.ExcelFile('2018_reg.xlsx')
df_2018_reg_1st = pd.read_excel(twenty18_reg, '1st')
df_2018_reg_2nd = pd.read_excel(twenty18_reg, '2nd')

In [19]:
df_2018_reg_1st

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0        Steven Adams        C   481/809  0.595  146/292  0.500     0  1108   
1       Jarrett Allen        C   335/568  0.590  197/278  0.709     6   873   
2       Deandre Ayton        C   509/870  0.585  141/189  0.746     0  1159   
3       Nicolas Batum       SF   253/562  0.450    77/89  0.865   116   699   
4     Darren Collison       PG   308/659  0.467  158/190  0.832    79   853   
5        Goran Dragic       PG   179/433  0.413   79/101  0.782    57   494   
6         Joel Embiid        C  580/1199  0.484  522/649  0.804    79  1761   
7       Tobias Harris       SF  611/1254  0.487  266/307  0.866   156  1644   
8          Chris Paul       PG   302/720  0.419  175/203  0.862   127   906   
9     Dennis Schroder       PG  457/1104  0.414  186/227  0.819   124  1224   
10       Jayson Tatum       SF  466/1036  0.450  195/228  0.855   116  1243   
11  Jonas Valanciunas        C   301/538  0.559  147/185  0.795    14   763   
12      Fred VanVleet       PG   246/600  0.410   97/115  0.843   112   701   

    REB  AST   ST  BLK   TO  auction_value  
0   760  124  117   76  135             13  
1   672  110   43  120  103             14  
2   729  125   61   67  126             20  
3   390  247   71   43  117              6  
4   232  459  110    9  125              6  
5   111  174   30    5   71             14  
6   871  234   46  122  226             41  
7   645  229   51   37  151             23  
8   265  473  114   18  152             38  
9   284  323   65   12  172             10  
10  477  168   84   57  122             10  
11  419   70   19   53   90              9  
12  167  307   57   20   82              0

In [20]:
df_2018_reg_2nd

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0     Thomas Bryant        C   309/502  0.616  107/137  0.781    33   758   
1       Mike Conley       PG  490/1120  0.438  343/406  0.845   155  1478   
2      Kevin Durant       SF  721/1383  0.521  448/506  0.885   137  2027   
3        Marc Gasol        C   390/870  0.448  192/253  0.759    99  1071   
4      Moe Harkless       SF   189/388  0.487    49/73  0.671    33   460   
5       Gary Harris       SG   270/637  0.424  115/144  0.799    82   737   
6        Joe Harris       SG   374/748  0.500  110/133  0.827   183  1041   
7       Jeremy Lamb       SG   431/979  0.440  231/260  0.888   115  1208   
8   Khris Middleton       SF  506/1148  0.441  216/258  0.837   179  1407   
9      Paul Millsap       PF   322/665  0.484  181/249  0.727    58   883   
10     Bobby Portis        C   279/628  0.444    77/97  0.794    75   710   
11  Josh Richardson       SF  423/1026  0.412  199/231  0.861   164  1209   
12   Andrew Wiggins       SF  498/1209  0.412  207/296  0.699   118  1321   

    REB  AST  ST  BLK   TO  auction_value  
0   454   92  25   67   60              0  
1   239  449  94   22  130             15  
2   497  457  58   84  225             65  
3   627  349  84   86  155             25  
4   269   74  67   53   45              0  
5   160  127  55   19   68             20  
6   291  181  38   17  121              0  
7   434  172  88   32   80              2  
8   461  331  80    7  174             32  
9   505  141  83   54   95             14  
10  403   72  35   20   74              4  
11  263  298  79   34  113             16  
12  352  184  71   48  138             11

In [21]:
print("Total value of 1st-place regular season team:", df_2018_reg_1st['auction_value'].sum())
print("Total value of 2nd-place regular season team:", df_2018_reg_2nd['auction_value'].sum())

Total value of 1st-place regular season team: 204
Total value of 2nd-place regular season team: 204


## 2018 - Playoffs

- Regular season 5th seed won 1st, Regular season 2nd seed won 2nd

In [22]:
twenty18_po = pd.ExcelFile('2018_playoff.xlsx')
df_2018_po_1st = pd.read_excel(twenty18_po, '1st')
df_2018_po_2nd = pd.read_excel(twenty18_po, '2nd')

In [23]:
df_2018_po_1st

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0      Patrick Beverley       PG   194/477  0.407   96/123  0.780   112   596   
1   Willie Cauley-Stein        C   412/741  0.556  140/254  0.551     1   965   
2        Andre Drummond        C  561/1052  0.533  243/412  0.590     5  1370   
3          De'Aaron Fox       PG  505/1102  0.458  303/417  0.727    86  1399   
4          Jerami Grant       PF   409/823  0.497  157/221  0.710   115  1090   
5            Kyle Lowry       PG   304/739  0.411  161/194  0.830   157   926   
6          JaVale McGee        C   400/641  0.624   96/152  0.632     1   897   
7       Larry Nance Jr.       PF   249/479  0.520   96/134  0.716    33   627   
8          Kelly Olynyk       PF   261/564  0.463  152/185  0.822   113   787   
9           Rajon Rondo       PG   175/432  0.405    23/36  0.639    51   424   
10     Tomas Satoransky       PG   257/530  0.485  131/160  0.819    64   709   
11    Russell Westbrook       PG  630/1473  0.428  296/451  0.656   119  1675   
12     Hassan Whiteside        C   388/680  0.571  109/243  0.449     2   887   

     REB  AST   ST  BLK   TO  auction_value  
0    388  300   67   43   85              2  
1    678  194   96   51   84              6  
2   1232  112  136  138  175             32  
3    304  590  133   45  227              7  
4    417   79   61  100   67              0  
5    312  564   91   31  182             31  
6    566   52   47  148  108              6  
7    552  214  100   40   97              5  
8    375  140   53   37  114              6  
9    243  367   57    7  127              1  
10   279  399   82   13  120              0  
11   807  784  142   33  325             57  
12   817   56   46  136   97             12

## Putting it all together (Last Part)

Combine all dataframes together and create visualizations around position + auction_value.

In [24]:
df_2015_po_1st = df_2015_reg_1st
df_2015_po_2nd = df_2015_reg_2nd

In [25]:
df_list = [df_2015_reg_1st, df_2015_reg_2nd, df_2015_po_1st, df_2015_po_2nd,
          df_2016_reg_1st, df_2016_reg_2nd, df_2016_po_1st, df_2016_po_2nd,
          df_2017_reg_1st, df_2017_reg_2nd, df_2017_po_1st, df_2017_po_2nd,
          df_2018_reg_1st, df_2018_reg_2nd, df_2018_po_1st, df_2018_po_2nd]

In [26]:
# print player names
# iterating through dictionary and checking condition (if column = player_name)

for df in df_list:
    for col in df:
        if col == 'player_name':
            print(df[col])

0          Nicolas Batum
1          Avery Bradley
2       Robert Covington
3          Jordan Farmar
4         Derrick Favors
5            Danny Green
6             Al Horford
7         Victor Oladipo
8          Elfrid Payton
9        Otto Porter Jr.
10       Josh Richardson
11    Karl-Anthony Towns
12          Kemba Walker
Name: player_name, dtype: object
0         Trevor Ariza
1          Jae Crowder
2        Toney Douglas
3            Pau Gasol
4          Rudy Gobert
5          George Hill
6         Shane Larkin
7           Kyle Lowry
8          Ian Mahinmi
9      Wesley Matthews
10     Thomas Robinson
11           Ish Smith
12    Hassan Whiteside
Name: player_name, dtype: object
0          Nicolas Batum
1          Avery Bradley
2       Robert Covington
3          Jordan Farmar
4         Derrick Favors
5            Danny Green
6             Al Horford
7         Victor Oladipo
8          Elfrid Payton
9        Otto Porter Jr.
10       Josh Richardson
11    Karl-Anthony Towns
12        

In [27]:
merged_df = pd.concat([df_2015_reg_1st, df_2015_reg_2nd, df_2015_po_1st, df_2015_po_2nd,
          df_2016_reg_1st, df_2016_reg_2nd, df_2016_po_1st, df_2016_po_2nd,
          df_2017_reg_1st, df_2017_reg_2nd, df_2017_po_1st, df_2017_po_2nd,
          df_2018_reg_1st, df_2018_reg_2nd, df_2018_po_1st, df_2018_po_2nd])

In [28]:
merged_df

player_name position     FGM/A    FG%    FTM/A    FT%  3PTM   PTS  \
0      Nicolas Batum       SF   372/873  0.426  163/192  0.849   139  1046   
1      Avery Bradley       SG  456/1019  0.447   96/123  0.780   147  1155   
2   Robert Covington       SF   274/712  0.385  140/177  0.791   170   858   
3      Jordan Farmar       PG    42/100  0.420    10/10  1.000    16   110   
4     Derrick Favors       PF   418/811  0.515  180/254  0.709     0  1016   
..               ...      ...       ...    ...      ...    ...   ...   ...   
8    Khris Middleton       SF  506/1148  0.441  216/258  0.837   179  1407   
9       Paul Millsap       PF   322/665  0.484  181/249  0.727    58   883   
10      Bobby Portis        C   279/628  0.444    77/97  0.794    75   710   
11   Josh Richardson       SF  423/1026  0.412  199/231  0.861   164  1209   
12    Andrew Wiggins       SF  498/1209  0.412  207/296  0.699   118  1321   

    REB  AST   ST  BLK   TO  auction_value  
0   428  403   65   43  205             17  
1   220  158  117   19  109              0  
2   419   97  105   40  143              8  
3    25   37   15    2   18              0  
4   503   94   74   91  104             16  
..  ...  ...  ...  ...  ...            ...  
8   461  331   80    7  174             32  
9   505  141   83   54   95             14  
10  403   72   35   20   74              4  
11  263  298   79   34  113             16  
12  352  184   71   48  138             11  

[208 rows x 14 columns]

In [29]:
merged_df['player_name'].value_counts()

Josh Richardson     5
Marc Gasol          5
Kemba Walker        5
Khris Middleton     4
Goran Dragic        4
                   ..
Luol Deng           1
Deandre Ayton       1
Isaiah Thomas       1
Tomas Satoransky    1
Jrue Holiday        1
Name: player_name, Length: 109, dtype: int64

In [30]:
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure
import numpy as np

plt.rc('xtick', labelsize=20) 
plt.rc('ytick', labelsize=20)
figure(num=None, figsize=(40, 10), dpi=150)
ax = pd.value_counts(merged_df['player_name']).plot.bar()
ax.set_ylabel("Frequency", fontsize=25)
ax.yaxis.set_label_coords(-.025,.5)

In [31]:
count = merged_df['player_name'].value_counts() 
print(count)

Josh Richardson     5
Marc Gasol          5
Kemba Walker        5
Khris Middleton     4
Goran Dragic        4
                   ..
Luol Deng           1
Deandre Ayton       1
Isaiah Thomas       1
Tomas Satoransky    1
Jrue Holiday        1
Name: player_name, Length: 109, dtype: int64


## Further ideas
- Where can I pay the least to get the most value?
- Does cost necessarily translate to value?
- What stage of the draft do most gems come out of?
- How about questions that people never ask? (I feel like the questions above are almost always answered. May the time (which 1/6th) of the draft can be important)

# Stage of Draft -- Analysis

Context: Auction drafts are not a sprint, like snake drafts. They can take a very long time. I am hoping to answer the following questions: 
- Does stage of draft matter? 
- What is the average cost at certain stages of the draft?
- What strategy can I take to grab some value?

Traditional 12-team league draft, 156 total picks. Divided into 6 "time chunk" draft portions.

In [32]:
### Import data

draft_res = pd.ExcelFile('auctiondraft_results.xlsx')
dr_2018 = pd.read_excel(draft_res, '2018')
dr_2017 = pd.read_excel(draft_res, '2017')
dr_2016 = pd.read_excel(draft_res, '2016')
dr_2015 = pd.read_excel(draft_res, '2015')

## Feature Creation -- Methodology

In a league with 12-teams and 13-player roster, there are 156 picks. I decided to segment the data into 6 sizeable chunks.
- Each chunk will contain 26 picks. We will assume each draft stage takes on average 30 minutes.
- In other words, by time the first 26 picks have completed, we would be 30 minutes *into* the draft.
- After 52 picks have completed, we would be 1 hour *into* the draft.

I will create a new column called **draft_stage**.

In [33]:
## Add new column with portion of draft
import numpy as np

dr_2018["draft_stage"] = np.nan
dr_2017["draft_stage"] = np.nan
dr_2016["draft_stage"] = np.nan
dr_2015["draft_stage"] = np.nan

dr_2016

Pick                             Player  Cost                 Team  \
0       1         Stephen Curry (GS - PG,SG)    65          Aminu Acids   
1       2         Dwight Howard (LAL - PF,C)    25  Ankle Insurance Pls   
2       3         Kevin Durant (Bkn - SF,PF)    66               Bidoof   
3       4              Chris Paul (OKC - PG)    53      fake basketball   
4       5        Draymond Green (GS - SF,PF)    38      Eamonn Shumpert   
..    ...                                ...   ...                  ...   
151   152          Dario Saric (Pho - SF,PF)     1          Aminu Acids   
152   153       Terrence Jones (Hou - SF,PF)     1               Bidoof   
153   154              Andrew Bogut (GS - C)     1   Wall About Benji's   
154   155         Emmanuel Mudiay (Uta - PG)     1   Wall About Benji's   
155   156  Michael Kidd-Gilchrist (Dal - SF)     1   Wall About Benji's   

     draft_stage  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
..           ...  
151          NaN  
152          NaN  
153          NaN  
154          NaN  
155          NaN  

[156 rows x 5 columns]

In [34]:
## Based on the pick #, assign the "stage" of the draft

def set_stage(df):

    for i in range(len(df)): 
        if df["Pick"][i] <=26: 
            df["draft_stage"][i] ="1st"

        elif df["Pick"][i] <= 52: 
            df["draft_stage"][i] ="2nd"

        elif df["Pick"][i] <= 78: 
            df["draft_stage"][i] ="3rd"

        elif df["Pick"][i] <= 104: 
            df["draft_stage"][i] ="4th"

        elif df["Pick"][i] <= 130: 
            df["draft_stage"][i] = "5th"

        else: 
            df["draft_stage"][i] = "6th"

set_stage(dr_2018)
set_stage(dr_2017)
set_stage(dr_2016)
set_stage(dr_2015)

## Add rankings, Merge dataframes from same year

Data sources:
- Draft Results from Yahoo! spanning from 2015-2018 
- Basketball Monster Player Rankings from 2015-2018 

Currently, the Draft Results data has the player's position in the **Player** field, we'll need to remove the paranthesis from this field.

In [35]:
dr_2018

Pick                            Player  Cost                 Team  \
0       1          Seth Curry (Dal - PG,SG)     1  I miss Marcus Smart   
1       2         Clint Capela (Atl - PF,C)    22    TYRONE LEWIS🏀�...   
2       3  Josh Richardson (Phi - PG,SG,SF)    16           DambuDanny   
3       4           Kevin Love (Cle - PF,C)    31      HospitAll-Stars   
4       5        Anthony Davis (LAL - PF,C)    73         Peter's Team   
..    ...                               ...   ...                  ...   
151   152        Ryan Anderson (Hou - PF,C)     1               Bidoof   
152   153          Jordan Bell (Mem - PF,C)     1      One Man's Trash   
153   154      Harry Giles III (Sac - PF,C)     1     Grand Theft Otto   
154   155             Alex Len (Sac - PF,C)     1               Bidoof   
155   156            Rajon Rondo (LAL - PG)     1     Grand Theft Otto   

    draft_stage  
0           1st  
1           1st  
2           1st  
3           1st  
4           1st  
..          ...  
151         6th  
152         6th  
153         6th  
154         6th  
155         6th  

[156 rows x 5 columns]

In [36]:
## Data cleaning

## Creates new 'Name' col that eliminates the 
## paranetheses in 'Player' col

def elim_par(df):
    df['Name'] = np.nan
    name_list = []

    for row in df['Player']:
        new_str = row.partition("(")[0]
        new_str = new_str.rstrip()
        name_list.append(new_str)

    df['Name'] = name_list

In [37]:
elim_par(dr_2018)
elim_par(dr_2017)
elim_par(dr_2016)
elim_par(dr_2015)

In [38]:
def round_average(df):
    
    count_1st = 0
    total_1st = 0 
    count_2nd = 0
    total_2nd = 0 
    count_3rd = 0
    total_3rd = 0 
    count_4th = 0
    total_4th = 0 
    count_5th = 0
    total_5th = 0 
    count_6th = 0
    total_6th = 0 

    for i in range(len(df)): 
        if df["draft_stage"][i] == '1st': 
            total_1st += df["Cost"][i] 
            count_1st += 1      

        elif df["draft_stage"][i] == '2nd': 
            total_2nd += df["Cost"][i] 
            count_2nd += 1 


        elif df["draft_stage"][i] == '3rd':  
            total_3rd += df["Cost"][i] 
            count_3rd += 1 


        elif df["draft_stage"][i] == '4th': 
            total_4th += df["Cost"][i] 
            count_4th += 1 


        elif df["draft_stage"][i] == '5th': 
            total_5th += df["Cost"][i] 
            count_5th += 1 

        else: 
            total_6th += df["Cost"][i] 
            count_6th += 1 

    print("Avg cost of player in 1st stage:", round(total_1st/count_1st))
    print("Avg cost of player in 2nd stage:", round(total_2nd/count_2nd))
    print("Avg cost of player in 3rd stage:", round(total_3rd/count_3rd))
    print("Avg cost of player in 4th stage:", round(total_4th/count_4th))
    print("Avg cost of player in 5th stage:", round(total_5th/count_5th))
    print("Avg cost of player in 6th stage:", round(total_6th/count_6th))

In [39]:
print("2018")
round_average(dr_2018)
print("\n2017")
round_average(dr_2017)
print("\n2016")
round_average(dr_2016)
print("\n2015")
round_average(dr_2015)

2018
Avg cost of player in 1st stage: 37.0
Avg cost of player in 2nd stage: 26.0
Avg cost of player in 3rd stage: 14.0
Avg cost of player in 4th stage: 8.0
Avg cost of player in 5th stage: 5.0
Avg cost of player in 6th stage: 1.0

2017
Avg cost of player in 1st stage: 41.0
Avg cost of player in 2nd stage: 23.0
Avg cost of player in 3rd stage: 12.0
Avg cost of player in 4th stage: 6.0
Avg cost of player in 5th stage: 5.0
Avg cost of player in 6th stage: 2.0

2016
Avg cost of player in 1st stage: 39.0
Avg cost of player in 2nd stage: 29.0
Avg cost of player in 3rd stage: 11.0
Avg cost of player in 4th stage: 7.0
Avg cost of player in 5th stage: 5.0
Avg cost of player in 6th stage: 1.0

2015
Avg cost of player in 1st stage: 43.0
Avg cost of player in 2nd stage: 22.0
Avg cost of player in 3rd stage: 13.0
Avg cost of player in 4th stage: 7.0
Avg cost of player in 5th stage: 4.0
Avg cost of player in 6th stage: 2.0


| stage| avg cost per player|
|------|----|
| 1st  |40  |
| 2nd  |25  |
| 3rd  |12.5|
| 4th  |7   |
| 5th  |4.75|
| 6th  |1.5 |

## Takeaways
- People tend to nominate more expensive players in the first stage (first 30 minutes) of a draft. In fact, people are more trigger-happy and loose with their budget in the beginning. A solid strategy would be to nominate overhyped, lower-ranked players in the earlier stages of the draft so you can get other managers to waste money, giving you more buying power as the draft progresses.
- The 3rd-5th stages offer a ton of valuable players for a small percentage of your budget. Stay alert during this timeframe!
 - ($12.5 of 200 = 6% of total budget, 7/200 = 3.5%, 4.75/200 = 2%)

## Rankings Analysis

- Read in BBM_PlayerRankings
- Merge dataframes together with draft results

In [40]:
bbm_pr = pd.ExcelFile('bbm_pr.xlsx')

bbm_pr_2018 = pd.read_excel(bbm_pr, '2018')
bbm_pr_2017 = pd.read_excel(bbm_pr, '2017')
bbm_pr_2016 = pd.read_excel(bbm_pr, '2016')
bbm_pr_2015 = pd.read_excel(bbm_pr, '2015')

In [41]:
bbm_pr_2018

Rank  Y!Adp9     Value                Name Team       Pos   g        m/g  \
0       1     1.7  1.267030        James Harden  HOU     PG/SG  78  36.752991   
1       2    13.3  0.915986         Paul George  OKC  SG/SF/PF  77  36.901732   
2       3     5.8  0.802711  Karl-Anthony Towns  MIN         C  77  33.048052   
3       4     6.1  0.746828        Kevin Durant  GSW     SF/PF  78  34.635256   
4       5    56.0  0.702161      Nikola Vucevic  ORL      PF/C  80  31.378958   
..    ...     ...       ...                 ...  ...       ...  ..        ...   
183   184   135.5 -0.444104        Bobby Portis  WAS      PF/C  50  25.964333   
184   185     0.0 -0.456472        Vince Carter  ATL     SG/SF  76  17.504167   
185   186    48.4 -0.462776         Gary Harris  DEN     SG/SF  57  28.758480   
186   187    81.7 -0.463984           Kris Dunn  CHI     PG/SG  46  30.205072   
187   188   136.4 -0.466001        Josh Jackson  PHO     SG/SF  79  25.166456   

           p/g       3/g        r/g       a/g       s/g       b/g       fg%  \
0    36.128205  4.846154   6.615385  7.512821  2.051282  0.730769  0.441592   
1    28.038961  3.792208   8.142857  4.129870  2.220779  0.441558  0.438042   
2    24.415584  1.844156  12.389610  3.363636  0.883117  1.623377  0.517871   
3    25.987179  1.756410   6.346154  5.858974  0.756410  1.076923  0.521330   
4    20.812500  1.050000  12.012500  3.837500  1.012500  1.137500  0.517725   
..         ...       ...        ...       ...       ...       ...       ...   
183  14.200000  1.500000   8.020000  1.440000  0.700000  0.420000  0.444268   
184   7.394737  1.618421   2.552632  1.144737  0.578947  0.355263  0.418803   
185  12.929825  1.438596   2.771930  2.228070  0.964912  0.333333  0.424528   
186  11.282609  0.739130   4.065217  6.021739  1.478261  0.456522  0.424901   
187  11.518987  0.924051   4.392405  2.316456  0.924051  0.708861  0.412604   

         fga/g       ft%      fta/g      to/g  
0    24.474359  0.878788  11.000000  4.961538  
1    20.961039  0.840445   7.000000  2.662338  
2    17.077922  0.835556   5.844156  3.129870  
3    17.730769  0.885375   6.487179  2.871795  
4    16.925000  0.788546   2.837500  1.987500  
..         ...       ...        ...       ...  
183  12.560000  0.793814   1.940000  1.480000  
184   6.157895  0.712121   0.868421  0.631579  
185  11.157895  0.798611   2.526316  1.175439  
186  11.000000  0.797101   1.500000  2.260870  
187  10.645570  0.671362   2.696203  2.189873  

[188 rows x 19 columns]

In [42]:
rankings_2018 = dr_2018.merge(bbm_pr_2018, on = 'Name')
rankings_2017 = dr_2017.merge(bbm_pr_2017, on = 'Name')
rankings_2016 = dr_2016.merge(bbm_pr_2016, on = 'Name')
rankings_2015 = dr_2015.merge(bbm_pr_2015, on = 'Name')

In [43]:
rankings_2015

Pick                           Player  Cost              Team_x  \
0       1       Anthony Davis (LAL - PF,C)    81      Anthony Harden   
1       2       Hassan Whiteside (Por - C)    25              Bidoof   
2       3       Stephen Curry (GS - PG,SG)    78  RileyCurry'sNaeNae   
3       4       James Harden (Hou - SG,SF)    75      Anthony Harden   
4       5            Chris Paul (OKC - PG)    62            TRILLSAP   
..    ...                              ...   ...                 ...   
119   151         Timofey Mozgov (Orl - C)     1      Anthony Harden   
120   152        P.J. Tucker (Hou - SG,SF)     1            TRILLSAP   
121   153     Andre Iguodala (Mia - SG,SF)     1       Jimmy Buckets   
122   154  Kristaps Porzingis (Dal - PF,C)     1      Anthony Harden   
123   156      Frank Kaminsky (Pho - PF,C)     1       Jimmy Buckets   

    draft_stage                Name  Rank  Y!Adp9     Value Team_y  ...  \
0           1st       Anthony Davis    17     1.0  0.409797    NOR  ...   
1           1st    Hassan Whiteside     9    33.0  0.610442    MIA  ...   
2           1st       Stephen Curry     1     2.0  1.493296    GSW  ...   
3           1st        James Harden     3     4.0  0.881308    HOU  ...   
4           1st          Chris Paul     7     6.0  0.670875    LAC  ...   
..          ...                 ...   ...     ...       ...    ...  ...   
119         6th      Timofey Mozgov   187   119.0 -0.422571    CLE  ...   
120         6th         P.J. Tucker    90   139.0 -0.137813    PHO  ...   
121         6th      Andre Iguodala   151   123.0 -0.322015    GSW  ...   
122         6th  Kristaps Porzingis    45   137.0  0.098210    NYK  ...   
123         6th      Frank Kaminsky   171   138.0 -0.385890    CHA  ...   

          3/g        r/g       a/g       s/g       b/g       fg%      fga/g  \
0    0.573770  10.278688  1.901639  1.278689  2.049180  0.492958  18.622951   
1    0.000000  11.849315  0.397260  0.602740  3.684932  0.605572   9.342466   
2    5.088608   5.443038  6.670886  2.139241  0.189873  0.503755  20.227848   
3    2.878049   6.109756  7.463415  1.695122  0.621951  0.439085  19.719512   
4    1.648649   4.189189  9.972973  2.054054  0.175676  0.462298  15.054054   
..        ...        ...       ...       ...       ...       ...        ...   
119  0.013158   4.434211  0.434211  0.289474  0.750000  0.565460   4.723684   
120  0.829268   6.243902  2.158537  1.292683  0.243902  0.410653   7.097561   
121  0.830769   4.046154  3.369231  1.123077  0.292308  0.478261   5.661538   
122  1.125000   7.305556  1.305556  0.736111  1.861111  0.420993  12.305556   
123  0.839506   4.135802  1.209877  0.469136  0.530864  0.409524   6.481481   

          ft%      fta/g      to/g  
0    0.758140   7.049180  1.983607  
1    0.650456   4.506849  1.876712  
2    0.907500   5.063291  3.316456  
3    0.860215  10.207317  4.560976  
4    0.896341   4.432432  2.621622  
..        ...        ...       ...  
119  0.715789   1.250000  0.934211  
120  0.746479   1.731707  1.353659  
121  0.614458   1.276923  1.215385  
122  0.837500   3.333333  1.680556  
123  0.729730   1.827160  0.716049  

[124 rows x 24 columns]

In [44]:
def rank_average(df):
    
    count_1st = 0
    total_1st = 0 
    count_2nd = 0
    total_2nd = 0 
    count_3rd = 0
    total_3rd = 0 
    count_4th = 0
    total_4th = 0 
    count_5th = 0
    total_5th = 0 
    count_6th = 0
    total_6th = 0 

    for i in range(len(df)): 
        if df["draft_stage"][i] == '1st': 
            total_1st += df["Rank"][i] 
            count_1st += 1      

        elif df["draft_stage"][i] == '2nd': 
            total_2nd += df["Rank"][i] 
            count_2nd += 1 


        elif df["draft_stage"][i] == '3rd':  
            total_3rd += df["Rank"][i] 
            count_3rd += 1 


        elif df["draft_stage"][i] == '4th': 
            total_4th += df["Rank"][i] 
            count_4th += 1 


        elif df["draft_stage"][i] == '5th': 
            total_5th += df["Rank"][i] 
            count_5th += 1 

        else: 
            total_6th += df["Rank"][i] 
            count_6th += 1 

    print("Avg rank of player in 1st stage:", round(total_1st/count_1st))
    print("Avg rank of player in 2nd stage:", round(total_2nd/count_2nd))
    print("Avg rank of player in 3rd stage:", round(total_3rd/count_3rd))
    print("Avg rank of player in 4th stage:", round(total_4th/count_4th))
    print("Avg rank of player in 5th stage:", round(total_5th/count_5th))
    print("Avg rank of player in 6th stage:", round(total_6th/count_6th))

In [45]:
rank_average(rankings_2018)

Avg rank of player in 1st stage: 31.0
Avg rank of player in 2nd stage: 65.0
Avg rank of player in 3rd stage: 74.0
Avg rank of player in 4th stage: 74.0
Avg rank of player in 5th stage: 105.0
Avg rank of player in 6th stage: 128.0


### Notes:
- no rank associated with 36 of the drafted players in 2018
- no rank associated with 43 of the drafted players in 2017
- no rank associated with 37 of the drafted players in 2016
- no rank associated with 32 of the drafted players in 2015

In [46]:
## LEFT JOIN

rankings_2018_lj = dr_2018.merge(bbm_pr_2018, on = 'Name', how='left')
rankings_2017_lj = dr_2017.merge(bbm_pr_2017, on = 'Name', how='left')
rankings_2016_lj = dr_2016.merge(bbm_pr_2016, on = 'Name', how='left')
rankings_2015_lj = dr_2015.merge(bbm_pr_2015, on = 'Name', how='left')

In [47]:
ranks_with_dr = pd.concat([rankings_2018_lj, rankings_2017_lj, rankings_2016_lj, rankings_2015_lj])

ranks_with_dr['Rank'] = ranks_with_dr['Rank'].fillna(value=189)
ranks_with_dr['Cost'] = ranks_with_dr['Cost'].fillna(value=0)

In [48]:
ranks_with_dr.groupby('Name')['Rank'].mean()

Name
Aaron Gordon          93.0
Al Horford            31.0
Al Jefferson         189.0
Al-Farouq Aminu      120.5
Alan Williams        189.0
                     ...  
Willy Hernangomez    189.0
Wilson Chandler      138.0
Zach LaVine          135.5
Zach Randolph        156.0
Zaza Pachulia        162.0
Name: Rank, Length: 251, dtype: float64

## Ideas for missing data...
- fill it in with a rank 189, then do mean and median for the draft stages

In [49]:
rankings_2018_lj['Rank'].fillna(value=189)

0      189.0
1       29.0
2       60.0
3      189.0
4       11.0
       ...  
151    189.0
152    189.0
153    189.0
154    162.0
155    189.0
Name: Rank, Length: 156, dtype: float64

In [50]:
rankings_2018_lj.groupby('draft_stage')['Rank'].mean()

draft_stage
1st     30.761905
2nd     65.478261
3rd     74.200000
4th     74.105263
5th    104.681818
6th    128.333333
Name: Rank, dtype: float64

## Observations

Interesting to see that the 3rd stage (1.5 hours in) and the 4th stage (2 hours in) are pretty close in median rank, but somewhat apart in cost.

Analysis on free agents? Total ranked players from free agency

The biggest challenge is finding data on actual, auction values

In [51]:
# Free agents that ranked in the top 100?

## Pull data from bbm_pr where adp is blank/null

# I feel like providing a summary simply based on frequency is not enough. Let me incorporate some rankings

In [52]:
# regular season 1st
final_2015_r1 = pd.merge(df_2015_reg_1st, bbm_pr_2015, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2015_r1 = pd.merge(final_2015_r1, dr_2015, on = 'Name', how = 'left')

final_2016_r1 = pd.merge(df_2016_reg_1st, bbm_pr_2016, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2016_r1 = pd.merge(final_2016_r1, dr_2016, on = 'Name', how = 'left')

final_2017_r1 = pd.merge(df_2017_reg_1st, bbm_pr_2017, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2017_r1 = pd.merge(final_2017_r1, dr_2017, on = 'Name', how = 'left')

final_2018_r1 = pd.merge(df_2018_reg_1st, bbm_pr_2018, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2018_r1 = pd.merge(final_2018_r1, dr_2018, on = 'Name', how = 'left')

# reg - 2nd
final_2015_r2 = pd.merge(df_2015_reg_2nd, bbm_pr_2015, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2015_r2 = pd.merge(final_2015_r2, dr_2015, on = 'Name', how = 'left')

final_2016_r2 = pd.merge(df_2016_reg_2nd, bbm_pr_2016, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2016_r2 = pd.merge(final_2016_r2, dr_2016, on = 'Name', how = 'left')

final_2017_r2 = pd.merge(df_2017_reg_2nd, bbm_pr_2017, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2017_r2 = pd.merge(final_2017_r2, dr_2017, on = 'Name', how = 'left')

final_2018_r2 = pd.merge(df_2018_reg_2nd, bbm_pr_2018, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2018_r2 = pd.merge(final_2018_r2, dr_2018, on = 'Name', how = 'left')

# playoffs 1st

final_2015_p1 = pd.merge(df_2015_po_1st, bbm_pr_2015, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2015_p1 = pd.merge(final_2015_p1, dr_2015, on = 'Name', how = 'left')

final_2016_p1 = pd.merge(df_2016_po_1st, bbm_pr_2016, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2016_p1 = pd.merge(final_2016_p1, dr_2016, on = 'Name', how = 'left')

final_2017_p1 = pd.merge(df_2017_po_1st, bbm_pr_2017, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2017_p1 = pd.merge(final_2017_p1, dr_2017, on = 'Name', how = 'left')

final_2018_p1 = pd.merge(df_2018_po_1st, bbm_pr_2018, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2018_p1 = pd.merge(final_2018_p1, dr_2018, on = 'Name', how = 'left')

# playoffs 2nd

final_2015_p2 = pd.merge(df_2015_po_2nd, bbm_pr_2015, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2015_p2 = pd.merge(final_2015_p2, dr_2015, on = 'Name', how = 'left')

final_2016_p2 = pd.merge(df_2016_po_2nd, bbm_pr_2016, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2016_p2 = pd.merge(final_2016_p2, dr_2016, on = 'Name', how = 'left')

final_2017_p2 = pd.merge(df_2017_po_2nd, bbm_pr_2017, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2017_p2 = pd.merge(final_2017_p2, dr_2017, on = 'Name', how = 'left')

final_2018_p2 = pd.merge(df_2018_po_2nd, bbm_pr_2018, left_on = 'player_name', right_on = 'Name', how = 'left')
final_2018_p2 = pd.merge(final_2018_p2, dr_2018, on = 'Name', how = 'left')

In [53]:
final_merge = pd.concat([final_2015_r1, final_2016_r1, final_2017_r1, final_2018_r1,
                        final_2015_r2, final_2016_r2, final_2017_r2, final_2018_r2,
                        final_2015_p1, final_2016_p1, final_2017_p1, final_2018_p1,
                        final_2015_p2, final_2016_p2, final_2017_p2, final_2018_p2])

In [54]:
final_merge['Rank'] = final_merge['Rank'].fillna(value=189)
final_merge['Cost'] = final_merge['Cost'].fillna(value=0)

In [55]:
pd.options.display.max_rows = 4000


final_merge.groupby('player_name')['Cost'].mean()

player_name
Aaron Gordon                 9.000000
Al Horford                  36.000000
Al-Farouq Aminu              2.000000
Andre Drummond              21.333333
Andrew Wiggins              11.000000
Anthony Davis               56.333333
Avery Bradley                2.333333
Bobby Portis                 4.000000
Carmelo Anthony             27.000000
Chris Paul                  45.500000
Damian Lillard              47.000000
Danny Green                 10.000000
Darren Collison              6.500000
De'Aaron Fox                 7.000000
DeMarcus Cousins            52.000000
Deandre Ayton               20.000000
Dennis Schroder             10.000000
Denzel Valentine             0.000000
Derrick Favors              16.000000
Dewayne Dedmon               0.000000
Dillon Brooks                0.000000
Dirk Nowitzki                4.000000
Donovan Mitchell             0.000000
Draymond Green              38.000000
Elfrid Payton                6.500000
Eric Bledsoe                22.000000


In [63]:
four_years_cost = ranks_with_dr.groupby('Name')['Cost'].median()
four_years_rank = ranks_with_dr.groupby('Name')['Rank'].median()


In [64]:
four_years_cost

Name
Aaron Gordon                 8.5
Al Horford                  27.0
Al Jefferson                21.0
Al-Farouq Aminu              2.0
Alan Williams                1.0
Alec Burks                   5.0
Alex Len                     1.0
Allen Crabbe                 3.5
Amir Johnson                 1.0
Andre Drummond              28.0
Andre Iguodala               1.0
Andrew Bogut                 1.0
Andrew Wiggins              11.5
Anthony Davis               65.5
Arron Afflalo                7.0
Avery Bradley                7.0
Bam Adebayo                  3.0
Ben McLemore                 1.0
Ben Simmons                 12.0
Bismack Biyombo              1.0
Blake Griffin               31.5
Bobby Portis                 2.5
Bogdan Bogdanovic            1.0
Bojan Bogdanovic             1.0
Bradley Beal                15.0
Brandon Ingram               4.5
Brandon Jennings             2.0
Brandon Knight              12.5
Brook Lopez                 15.0
Buddy Hield                  3.0
C.J. 

In [64]:
four_years_stats = pd.merge(four_years_cost, four_years_rank, on = 'Name', how = 'left')

In [65]:
four_years_stats.sort_values('Rank')

Cost   Rank
Name                                 
Karl-Anthony Towns        61.5    2.5
James Harden              71.5    3.5
Kevin Durant              66.5    4.5
Stephen Curry             65.5    4.5
Anthony Davis             65.5    7.0
Giannis Antetokounmpo     63.5    7.5
Nikola Jokic              46.0    8.0
Russell Westbrook         64.5   10.0
Paul George               44.0   11.0
LeBron James              58.5   11.5
Kawhi Leonard             53.5   12.5
Damian Lillard            46.5   13.0
Kemba Walker              31.5   16.0
LaMarcus Aldridge         29.5   21.0
Isaiah Thomas             22.0   23.0
Jimmy Butler              37.0   23.5
Chris Paul                49.0   23.5
Klay Thompson             29.5   24.0
Bradley Beal              15.0   24.0
Kyrie Irving              35.0   25.5
Draymond Green            34.5   25.5
Kyle Lowry                33.0   26.0
Clint Capela              11.0   29.0
Brook Lopez               15.0   30.5
Marc Gasol                27.0   33.5
Khris Middleton           14.0   34.5
DeMar DeRozan             18.5   34.5
Tobias Harris             15.5   35.0
Andre Drummond            28.0   35.5
Al Horford                27.0   36.0
Deandre Ayton             20.0   36.0
Gorgui Dieng               8.0   36.0
Rudy Gobert               28.5   37.0
Jeremy Lamb                2.0   38.0
Jayson Tatum              10.0   40.0
Eric Bledsoe              22.0   41.0
Donovan Mitchell          36.0   42.0
Jrue Holiday               9.5   42.5
Thaddeus Young             5.0   42.5
Nikola Vucevic            15.5   43.0
Montrezl Harrell           3.0   43.0
Ben Simmons               12.0   44.0
DeMarcus Cousins          53.5   44.5
JaVale McGee               6.0   45.0
Joel Embiid               40.0   45.0
Myles Turner              25.0   45.0
Kristaps Porzingis        25.0   45.0
Jae Crowder                4.0   46.0
Josh Richardson           12.0   46.0
Trevor Ariza              13.5   48.0
Serge Ibaka               19.5   49.0
Ricky Rubio                9.0   50.0
Jamal Murray               9.5   50.5
Dirk Nowitzki              8.0   52.0
Steven Adams               4.0   52.0
DeAndre Jordan            24.0   52.5
Kevin Love                27.5   53.5
Hassan Whiteside          31.5   54.0
Jarrett Allen             14.0   55.0
Lou Williams               8.0   55.5
Jeff Teague               12.5   56.5
Joe Ingles                11.0   57.0
Robert Covington           8.5   57.5
Derrick Favors            12.5   61.5
Bojan Bogdanovic           1.0   62.0
Darren Collison            5.5   62.0
Buddy Hield                3.0   63.0
Jonas Valanciunas          9.5   64.0
Paul Millsap              30.0   65.0
Greg Monroe                8.0   65.0
Victor Oladipo            27.0   67.0
Mike Conley               21.0   69.0
Marvin Williams            4.5   70.5
Carmelo Anthony           23.0   70.5
Nicolas Batum             11.5   71.0
Dewayne Dedmon             2.0   75.0
Courtney Lee               1.0   75.0
Gordon Hayward            22.5   76.0
Amir Johnson               1.0   77.0
Kentavious Caldwell-Pope   2.5   77.5
Bam Adebayo                3.0   79.0
Trae Young                 8.0   81.0
John Collins              14.0   85.0
Chris Bosh                23.0   85.0
Mason Plumlee              3.0   87.0
Danilo Gallinari           8.5   88.0
Enes Kanter                7.5   89.0
Robin Lopez                4.0   89.0
Pau Gasol                 11.5   89.0
Blake Griffin             31.5   89.5
P.J. Tucker                1.0   90.0
Jonathan Isaac             6.0   90.0
Marcin Gortat              6.5   90.0
Aaron Gordon               8.5   90.5
Harrison Barnes            7.0   90.5
Dwight Howard             14.0   91.0
Larry Nance Jr.            3.0   92.5
Goran Dragic              16.0   93.0
Wesley Matthews            3.0   93.0
Kelly Oubre Jr.            1.0   93.0
Luka Doncic               16.0   95.0
Andrew Wiggins            11.5   95.0
Domantas Sabonis           2.0   96.0
Gary Harris               11.0   99.

In [66]:
four_years_stats.sort_values(by=['Cost', 'Rank'])

Cost   Rank
Name                                 
Bojan Bogdanovic           1.0   62.0
Courtney Lee               1.0   75.0
Amir Johnson               1.0   77.0
P.J. Tucker                1.0   90.0
Kelly Oubre Jr.            1.0   93.0
Jordan Clarkson            1.0  109.0
Andre Iguodala             1.0  110.0
Tyler Johnson              1.0  114.0
Bogdan Bogdanovic          1.0  116.0
Terry Rozier               1.0  130.0
Langston Galloway          1.0  133.0
Andrew Bogut               1.0  134.0
Isaiah Canaan              1.0  165.0
Alex Len                   1.0  166.0
Frank Kaminsky             1.0  166.5
Collin Sexton              1.0  174.0
Josh Hart                  1.0  175.0
C.J. Miles                 1.0  183.0
Kris Dunn                  1.0  187.0
Timofey Mozgov             1.0  188.0
Josh Jackson               1.0  188.5
Alan Williams              1.0  189.0
Ben McLemore               1.0  189.0
Bismack Biyombo            1.0  189.0
Dante Exum                 1.0  189.0
Devin Harris               1.0  189.0
Doug McDermott             1.0  189.0
Harry Giles III            1.0  189.0
Jared Sullinger            1.0  189.0
Joffrey Lauvergne          1.0  189.0
Jordan Bell                1.0  189.0
Justise Winslow            1.0  189.0
Malik Monk                 1.0  189.0
Marcus Morris Sr.          1.0  189.0
Nick Young                 1.0  189.0
Sergio Rodriguez           1.0  189.0
Seth Curry                 1.0  189.0
Trevor Booker              1.0  189.0
Taj Gibson                 1.5  105.5
Michael Kidd-Gilchrist     1.5  121.5
Ersan Ilyasova             1.5  130.5
Ed Davis                   1.5  140.5
Jeff Green                 1.5  160.5
Jeremy Lamb                2.0   38.0
Dewayne Dedmon             2.0   75.0
Domantas Sabonis           2.0   96.0
Patrick Beverley           2.0   99.0
Joe Johnson                2.0  115.0
Al-Farouq Aminu            2.0  120.5
J.R. Smith                 2.0  130.0
Roy Hibbert                2.0  131.0
Wilson Chandler            2.0  134.0
Miles Bridges              2.0  138.0
Will Barton                2.0  158.0
Tony Parker                2.0  164.5
Jakob Poeltl               2.0  168.0
Brandon Jennings           2.0  189.0
David Lee                  2.0  189.0
Mo Williams                2.0  189.0
Tim Frazier                2.0  189.0
Kentavious Caldwell-Pope   2.5   77.5
Luol Deng                  2.5  127.0
Tristan Thompson           2.5  146.5
JaMychal Green             2.5  148.0
Bobby Portis               2.5  186.5
Montrezl Harrell           3.0   43.0
Buddy Hield                3.0   63.0
Bam Adebayo                3.0   79.0
Mason Plumlee              3.0   87.0
Larry Nance Jr.            3.0   92.5
Wesley Matthews            3.0   93.0
Kent Bazemore              3.0  129.0
Cedi Osman                 3.0  133.0
Malcolm Brogdon            3.0  134.5
Rajon Rondo                3.0  137.0
Ish Smith                  3.0  145.0
Deron Williams             3.0  158.5
Zaza Pachulia              3.0  162.0
Tyson Chandler             3.0  164.0
Marvin Bagley III          3.0  165.0
Marcus Smart               3.0  176.0
Caris LeVert               3.0  189.0
Kevin Knox II              3.0  189.0
Kevin Martin               3.0  189.0
Nik Stauskas               3.0  189.0
Skal Labissiere            3.0  189.0
Kelly Olynyk               3.5  100.0
T.J. Warren                3.5  122.5
Allen Crabbe               3.5  134.0
Evan Turner                3.5  141.5
Michael Carter-Williams    3.5  183.0
Jae Crowder                4.0   46.0
Steven Adams               4.0   52.0
Robin Lopez                4.0   89.0
De'Aaron Fox               4.0  114.0
Dario Saric                4.0  128.0
Jaylen Brown               4.0  152.0
Derrick Rose               4.0  189.0
Emmanuel Mudiay            4.0  189.0
Joakim Noah                4.0  189.0
Markelle Fultz             4.0  189.0
Marquese Chriss            4.0  189.0
Terrence Jones             4.0  189.0
Trey Burke                 4.0  189.

# Winners


| Player Name | Frequency | Avg. Rank (on winning roster) | Avg. Rank (across 4 years) | Avg. Cost (on winning roster) | Avg. Cost (across 4 years)|
| ----------- | --------- | ----------------------------- | -------------------------- | ----------------------------- | ------------------------- |
| Josh Richardson |          5 | 136 | 46 | 6.3 | 12 |
| Kemba Walker    |          5 | 17.7 | 17.3 | 24.3 | 29 |
| Marc Gasol       |           5| 32.4 | 50.5 | 25 | 25.8 |
| Kyle Lowry     |             4 | 27.3 | 28.5 | 34 | 32.5 |
| Khris Middleton  |        4 | 29.5 | 67.5 | 25.5 | 15.5 |
| Pau Gasol        |           4|51|96.3|17.5|14.3|
| Goran Dragic   |            4|100.3|106.5|11.8|15.3|
| Elfrid Payton  |          4|160|143|6.5|6|
| Avery Bradley              | 3|73.7|163|2.3|6|
| Nikola Vucevic             | 3|52|37|16|19.5|
| Kevin Durant                |3|5.3|4.7|65.7|67|
| Thaddeus Young             |3|49.3|47|6.3|6|
| Dirk Nowitzki              | 3|83.3|76.3|4|7.3|
| Jae Crowder                |3|38.7|80.3|4|4.7|
| Jarrett Allen              |3|105.7|55|4.7|14|
| Robert Covington           |3|62.3|72.5|6.3|8.3|
| Trevor Ariza               |3|28|52.3|16.7|12.5|
| Andre Drummond             |3|15.7|45.3|21.3|27|
| Anthony Davis              |3|1.7|8|56.3|66.3|
| Russell Westbrook          |3|18.7|13.3|63|64.5|
| Nicolas Batum              |3|66.7|77.8|13.3|11.5|
| Mike Conley                |3|27.7|88.5|15.7|21.8|
| Hassan Whiteside           |3|38.7|53.8|20.7|31.8|